File: Beecrowd_Parte5/Socrates_e_Suas_perguntas_3155.c
#include <stdio.h>
#include <stdlib.h>

// Estrutura para armazenar uma consulta
typedef struct {
    int L, R, index;
} Query;

// Função de comparação para ordenar as consultas por blocos
int compareQueries(const void *a, const void *b) {
    Query *query1 = (Query *)a;
    Query *query2 = (Query *)b;
    
    // Ordena por blocos para minimizar a movimentação no array
    int blockSize = 500; // Tamanho do bloco ajustável
    int block1 = query1->L / blockSize;
    int block2 = query2->L / blockSize;
    
    if (block1 != block2) {
        return block1 - block2;
    } else {
        return query1->R - query2->R;
    }
}

int main() {
    int N, Q;
    scanf("%d %d", &N, &Q);

    int array[N + 1];
    for (int i = 1; i <= N; i++) {
        scanf("%d", &array[i]);
    }

    // Leitura das consultas e armazenamento em um array de estruturas Query
    Query queries[Q];
    for (int i = 0; i < Q; i++) {
        scanf("%d %d", &queries[i].L, &queries[i].R);
        queries[i].index = i;
    }

    // Ordenar as consultas usando a função de comparação
    qsort(queries, Q, sizeof(Query), compareQueries);

    // Arrays para contagem de frequência
    int count[N + 1];
    int freq[N + 1];
    int answer[Q];

    // Inicialização dos arrays de contagem
    for (int i = 0; i <= N; i++) {
        count[i] = 0;
        freq[i] = 0;
    }

    int currentL = 1, currentR = 0;
    int distinctCount = 0;

    // Função para adicionar um elemento no intervalo
    auto add = [&](int num) {
        int currentCount = count[num];
        int currentFreq = freq[currentCount];
        int newCount = currentCount + 1;
        int newFreq = freq[newCount];

        freq[currentCount]--;
        freq[newCount]++;
        count[num] = newCount;

        if (newCount == num) {
            distinctCount++;
        }
        if (currentCount == num) {
            distinctCount--;
        }
    };

    // Função para remover um elemento do intervalo
    auto remove = [&](int num) {
        int currentCount = count[num];
        int currentFreq = freq[currentCount];
        int newCount = currentCount - 1;
        int newFreq = freq[newCount];

        freq[currentCount]--;
        freq[newCount]++;
        count[num] = newCount;

        if (newCount == num) {
            distinctCount++;
        }
        if (currentCount == num) {
            distinctCount--;
        }
    };

    // Processamento das consultas
    for (int i = 0; i < Q; i++) {
        int L = queries[i].L;
        int R = queries[i].R;
        int index = queries[i].index;

        // Ajustar o intervalo para [L, R]
        while (currentR < R) {
            currentR++;
            add(array[currentR]);
        }
        while (currentR > R) {
            remove(array[currentR]);
            currentR--;
        }
        while (currentL < L) {
            remove(array[currentL]);
            currentL++;
        }
        while (currentL > L) {
            currentL--;
            add(array[currentL]);
        }

        // Calcular a resposta para a consulta atual
        answer[index] = distinctCount;
    }

    // Imprimir as respostas
    for (int i = 0; i < Q; i++) {
        printf("%d\n", answer[i]);
    }

    return 0;
}

File: Beecrowd_Parte5/Socrates_e_Suas_perguntas_3155.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Socrates_e_Suas_perguntas_3155 {

    static class Query {
        int L, R, index;

        public Query(int L, int R, int index) {
            this.L = L;
            this.R = R;
            this.index = index;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());

        int N = Integer.parseInt(tokenizer.nextToken());
        int Q = Integer.parseInt(tokenizer.nextToken());

        int[] array = new int[N + 1];
        tokenizer = new StringTokenizer(reader.readLine());
        for (int i = 1; i <= N; i++) {
            array[i] = Integer.parseInt(tokenizer.nextToken());
        }

        Query[] queries = new Query[Q];
        for (int i = 0; i < Q; i++) {
            tokenizer = new StringTokenizer(reader.readLine());
            int L = Integer.parseInt(tokenizer.nextToken());
            int R = Integer.parseInt(tokenizer.nextToken());
            queries[i] = new Query(L, R, i);
        }

        // Ordenar as queries por blocos para minimizar o movimento no array
        int blockSize = (int) Math.sqrt(N);
        Arrays.sort(queries, Comparator.comparingInt((Query q) -> q.L / blockSize)
                                      .thenComparingInt(q -> q.R));

        int[] count = new int[N + 1];
        int[] freq = new int[N + 1];
        int[] answer = new int[Q];

        int currentL = 1, currentR = 0;
        int distinctCount = 0;

        // Função para adicionar um elemento no intervalo
        Runnable add = () -> {
            int num = array[currentR];
            freq[count[num]]--;
            count[num]++;
            freq[count[num]]++;
            if (count[num] == num) {
                distinctCount++;
            }
            if (count[num] - 1 == num) {
                distinctCount--;
            }
        };

        // Função para remover um elemento do intervalo
        Runnable remove = () -> {
            int num = array[currentL];
            freq[count[num]]--;
            if (count[num] == num) {
                distinctCount--;
            }
            count[num]--;
            freq[count[num]]++;
            if (count[num] == num) {
                distinctCount++;
            }
        };

        for (Query query : queries) {
            int L = query.L, R = query.R, idx = query.index;

            while (currentR < R) {
                currentR++;
                add.run();
            }
            while (currentR > R) {
                remove.run();
                currentR--;
            }
            while (currentL < L) {
                remove.run();
                currentL++;
            }
            while (currentL > L) {
                currentL--;
                add.run();
            }

            answer[idx] = distinctCount;
        }

        StringBuilder output = new StringBuilder();
        for (int ans : answer) {
            output.append(ans).append("\n");
        }
        System.out.print(output);
    }
}
